### PR TITLE
dzg workaround

### PR DESCRIPTION
**Pull request by jmberg**
This implements the dzg workaround - first patch is a cleanup to remove some duplicated code, and then the second tries to detect DZG meters with their broken reporting of positive numbers.

### CODE REVIEW

**Comment by jmberg** [comment]
and pretty please don't squash these changes _if you decide to merge them at all_ ... I have a rather explanatory commit log that would be good to include in the history

File: sml/src/sml_list.c [hunk 1]
```diff
@@ -147,12 +156,57 @@ sml_list *sml_list_entry_parse(sml_buffer *buf) {
 	l->scaler = sml_i8_parse(buf);
 	if (sml_buf_has_errors(buf)) goto error;
 
+	value_tl = sml_buf_get_current_byte(buf);
```
Comment (mbehr1) [inline sml/src/sml_list.c]: just a comment (was part of prev. code already): l doesn't get freed (sml_list_free(l)) if any of the prev. errors occurs.

**Comment by jmberg** [comment]
Frankly, this whole code is a mess regardless of how the lists are handled. The whole list allocation and all that is just completely unnecessary - it's entirely possible to parse this stuff without ever doing any memory allocation at all, just by keeping appropriate pointers/offsets into the message structure. You could just do the list iteration from the original message data, at least once you've parsed it once to understand that it's well-formed to start with...

I was sorely tempted to just make a new library to parse all this, and I may still do that as an excuse to learn something like rust, but for now I also wanted my meter to work, hence this workaround :-)

**Comment by andig** [comment]
OT

> I was sorely tempted to just make a new library to parse all this, and I may still do that as an excuse to learn something like rust, but for now I also wanted my meter to work, hence this workaround :-)

If anybody wants it I have started a go port of this which is a language I've really come to like and which excels at concurrency...

**Pull request merged**